Compilation for Linux, MacOS, and Windows

kuniss opened the pull request:
- applied matrix compilation and testing as suggested by the [GitHub action for setting up dlang](https://github.com/dlang-community/setup-dlang)
- MacOS compilation and testing worked out of the box, Windows alternative needed some adaptations...
- adjustments to get internal D compiler invocation for generated compiler to run on Windows
- enhancements to test helper functions to get tests passed on Windows 
- adjustments to some tests where Windows command line is completely different to Unix-like OS
- renamed GitHub action build file _d.yml_ to _build.yml_ as later we will add a _release.yml_ file

An example of an successful run may be found at [here](https://github.com/kuniss/gamma/actions/runs/4337921745).

@linkrope, I created a PR to allow you to check style of the D source code adaptations, I did and to take the advantages of the GitHub PR tooling.

kuniss committed:
Added Windows specific modifications

of the DMD compiler imvocation to enable compilation on Windows.

kuniss committed:
Added Windows specifc modidications and helper function

to get tests passed on Windows.
Added Windows specific helper function asSingleLineDosInput
is for inline EAG source code tests under Windows.

kuniss committed:
Added Windows specific test code alternatives

as Windows is very different here to Linux and MacOS and
any Unix-kind OS.

kuniss committed:
Enabled matrix build for Linux, MaxOS, and Windows

and the according tests.

kuniss committed:
Renamed build script to "build.yml"

as later we want to add a "release.yml".

kuniss committed:
Removed unittest job

as already contained in example tests.

kuniss committed:
Switched to ldc-latest for faster code.

kuniss committed:
Swicthed back to DMD for testing.

kuniss commented:
@linkrope Merry Christmas! :smile: 
Did you see this PR?
I would add to it the executable publication as asset. 
Probably I will use https://github.com/marketplace/actions/upload-assets-to-a-release for the publication...

kuniss committed:
Added a release worflow

which adds binaries to a published GitHub release.

kuniss committed:
Pinned dmd and ldc versions

as "latest" fails. linkrope did it too for dmd.
The versino of ldc was looked up at https://github.com/ldc-developers/ldc/releases matching the version of Frontend, druntime and Phobos to that of dmd.

kuniss committed:
Pinned dmd and ldc versions

as "latest" fails. linkrope did it too for dmd.
The versino of ldc was looked up at https://github.com/ldc-developers/ldc/releases matching the version of Frontend, druntime and Phobos to that of dmd.

kuniss committed:
Give executables unique names before upload

kuniss committed:
Avoid GH releases are triggering build.yml

by ignoring all tags pushes as trigger.

kuniss committed:
Corrected release exe renaming for Windows

as it is a power shell running by default, not a dos shell.

kuniss committed:
Ignore tags starting with 'v' only

assuming those are the tags from a release publication. Otherwise, with '*' pattern, build.yml is not triggered at all.

kuniss committed:
One asterisk is enough for matching a tag

to be ignored according to https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-branches-and-tags

kuniss committed:
Force a push to trigger a build

get missed before. Maybe, another order works better...

kuniss committed:
Revert back to original push event specification

as excluding version tags let any git push skip the build. Not clear why.

kuniss committed:
Ensure test is running before release upload.

kuniss commented:
Seems, I got the publishing of the 3 different gamma executables working too.
It builds on my fork as visible in [this build](https://github.com/kuniss/gamma/actions/runs/7342567290).
And it gets published to the [release page which triggered the build](https://github.com/kuniss/gamma/releases/tag/v0.3.0-beta-4).

The only thing, I did not get managed is to avoid duplicate on tag pushing. It does not work as described at the GitHub Actions documentation...

kuniss committed:
Upgraded dscanner to version 0.14.0

as in master, see ci.yml.

linkrope commented on src/gamma/main.d:
```diff
@@ -239,9 +239,28 @@ void build(string[] fileNames, string outputDirectory)
+        args ~= format!"-of=%s"(executableName);
     }
+
+    version(Windows) 
```
Isn't this already fixed in `test/helper.d`?
On Windows, gamma should be called with paths `like\this` instead of `like/this`.
(You don't translate `outputDirectory` either)

kuniss replied:
As far as I remember (the PR is from March, so I'm not completely sure anymore :-) ), this code is not  because of the user input, but because of the include files listed some lines above at https://github.com/linkrope/gamma/blob/11c93a9229e028f778573de87b4f478dda96f508/src/gamma/main.d#L238

I think, as long as we have this includes (I remember, you mentioned, we may get rid of it through a library), this is the best way of handling the different paths on Windows. What do you think?

linkrope replied:
Replace literal string with calls to `buildPath`?
https://dlang.org/phobos/std_path.html#buildPath

    ~ buildPath("src", "io.d") ~ buildPath("src", "log.d") ~ buildPath("src", "epsilon", "soag", "listacks.d");

kuniss replied:
Oh, thanks for the hint! That was beyond may D-knowledge, yet.

kuniss replied:
Implemented. Please review.

kuniss replied:
No! I mixed this up with another change. This is not implemented yet!

kuniss replied:
Done with de88bc1c7f4c9bba260f1166044a9e752cc4901a. :smile:

kuniss replied:
BTW, is there a way to incorporate these runtime sources into the gamma executable, @linkrope?
That would make the gamma executable working standalone, compiling specified compilers to an executable as long as a DMD is installed.

linkrope replied:
gamma should just generate a dub.json file for the generated compiler.
First problem is that there is already a dub.json file in the directory checked out:
the one for the gamma generator...

kuniss replied:
In my VScode extension I generate the compiler source code into a configurable build directory (by default `build`, but could also be `gen` or so). Maybe we should go this way in the gamma project too? Maybe we sould always create there a subdirectory with the compilers name, in case some one has 2 EAG specifications in one project.
Then the generated compilers' _dub.json_ could be generated there.

But how to deal with the needed include files?

linkrope commented on src/gamma/main.d:
```diff
@@ -239,9 +239,28 @@ void build(string[] fileNames, string outputDirectory)
 
     if (!outputDirectory.empty)
     {
+        version(Windows)
```
I would prefer such a helper function:
https://github.com/dlang/dmd/blob/master/compiler/src/build.d#L1729

    args ~= format!"-of=%s"(fileNames.front.stripExtension.executableName);

kuniss replied:
I will do so. I'm I implemented it already on another branch and it works well.

kuniss replied:
Done.
Do you want to resolve the conversation? It will show on GH that it has been done...

linkrope commented on .github/workflows/d.yml:
Then, you must also fix the badge in the README?
https://github.com/linkrope/gamma/blob/master/README.md?plain=1#L5

(Or, rename build.yml into ci.yml?)

kuniss replied:
I guess, I will rename it to ci.yml.

kuniss replied:
I renamed it, but it is in conflict with your ci.yml. So, it builds only on my fork: https://github.com/kuniss/gamma/actions/runs/7365335289

kuniss replied:
Thanks for the merge!

linkrope commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,66 @@
+
+jobs:
+
+  test:
```
Avoid copy&paste: remove test job, assuming it's already run on push?

kuniss replied:
This wasn't a copy'n'paste error but I intended! It proves that the commit going to be released passes all tests, to prevent mistakes...

linkrope replied:
I did not imply a copy&paste error. I was just complaining about copy&paste at all.
As a result, it becomes tedious to maintain the test job twice.

Either you find out how to provide jobs in one place and include them. Or you do without the test job during the release, because it must have run during the previous commit anyway.

kuniss replied:
Ok, I understood your point. I guess, I have idea how to achieve this.

kuniss committed:
Added executableName() for Windows' ".exe" extension

as suggested by @linkrope and using it build().

kuniss committed:
Renamed build.yml to ci.yml

as discussed on PR here: https://github.com/linkrope/gamma/pull/17/files/11c93a9229e028f778573de87b4f478dda96f508#r1437757527

linkrope committed:
Merge branch 'master' into cross-compilation

kuniss committed:
Refactored out reusable workflows

from ci.yml, to refernce them later there and in release.yml

kuniss committed:
Referenced reusable workflows

kuniss committed:
Define compiler versions as workflow parameters

kuniss committed:
Added uploading of built executables

kuniss committed:
Corrected source path for upload step

Misused due to a misunderstanding.

kuniss committed:
Bumped actions/upload-artifact to v4

kuniss committed:
Rewritten release workflow

based on reusable workflows used at ci.yml and passing built binaries from build job using download-artifact action.

kuniss committed:
Corrected the renaming for the Windows binary

as it all happens on Linux, no Windows runner involved anymore here as in the first workflow version.

kuniss committed:
Avoid CI workflow is running on release publications

and within twice, once for the tag pushed for release and second for the release event.
This was found in https://stackoverflow.com/questions/70743715/how-do-i-configure-a-github-actions-workflow-so-it-does-not-run-on-a-tag-push

kuniss committed:
Improved matrix names in GH Actions GUI

kuniss committed:
Improved formating

kuniss committed:
Refactored out upload workflow

into upload.yml to have the same abstractions levels for jobs in release.yml

kuniss committed:
Removed secret declaration for GITHUB_TOKEN

as it shows an operational error on push. Seems, this token is implicitely inherited.

kuniss commented:
I guess, I found a good solution now. Please review it.

The release binary upload to the release page is now done using the upload/download GH Actions, as it is intended, I guess.
For the result see
- a CI build: https://github.com/linkrope/gamma/actions/runs/7415578645
- a release build: https://github.com/kuniss/gamma/actions/runs/7415515452
- a resulting release page with binaries: https://github.com/kuniss/gamma/releases/tag/v0.3.0-gamma.3 

I set the uploaded artifact redemption period for a build to 1 day, as it is mainly needed for upload when a release is published. But we could extend it to 90 days max. What do you think?
On a release page the uploaded artifacts should stay forever...

kuniss committed:
Get rid of OS specific file path separators

as suggested by @linkrope in https://github.com/linkrope/gamma/pull/17#discussion_r1438677466.
Not tested yet on Windows as locally not available, currently. Runs wel on Linux.

linkrope commented on src/gamma/main.d:
```diff
@@ -250,3 +254,11 @@ void build(string[] fileNames, string outputDirectory)
         exit(status);
 }
+
+string executableName(const string name)
```
Make it explicitly `private`:

    private string executableName(const string name)

kuniss replied:
For my curiosity, why only this method, and not the others in _main.d_ too?
I guess, I "copied" it from the other methods above.

kuniss replied:
Done with 49328b890081262840a0dd7cace4f266e74dbee1.

linkrope commented on src/gamma/main.d:
```diff
@@ -250,3 +254,11 @@ void build(string[] fileNames, string outputDirectory)
+        return name ~ ".exe";
+    else
+        return name;
+}
```
Get rid of this red "minus in circle": in VS Code, enable "insertFinalNewline"

kuniss replied:
I do not see the red circle well in my light theme. 😉

kuniss replied:
Thank for the configuration hint regarding "insertFinalNewline"! Didn't know it.

kuniss replied:
Done with 49328b890081262840a0dd7cace4f266e74dbee1.

kuniss committed:
Made locally only used method private

as suggested by @linkrope.

kuniss commented:
@linkrope, I just encountered, that `dub build` is building a debug version:
```
denis@lisp:~/git/gamma$ dub build
    Starting Performing "debug" build using /usr/bin/dmd for x86_64.
``` 
Is my interpretation right?
If so, shouldn't we build a non-debug version when releasing?

linkrope commented:
> @linkrope, I just encountered, that `dub build` is building a debug version:
> 
> ```
> denis@lisp:~/git/gamma$ dub build
>     Starting Performing "debug" build using /usr/bin/dmd for x86_64.
> ```
> 
> Is my interpretation right? If so, shouldn't we build a non-debug version when releasing?

See: https://dub.pm/dub-reference/buildtypes/#default-examples

("optimize", "inline", debugInfo") would be nice: fast code, still with helpful diagnostics.

Note that array bounds check is still enabled in release mode. There is a special build type for release builds with no bounds check. Similiarly, if you believe that contracts (pre -conditions, post-conditstions, invariants) are as valuable as bounds checks, then you should avoid release mode, where contracts are skipped. And use just "optimize" and "inline" instead. The "debug" is not about speed but about size: wouldn't you like to have debug symbols available to understand a backtrace in case of a fatal error?

kuniss commented:
> ("optimize", "inline", debugInfo") would be nice: fast code, still with helpful diagnostics.
> 
> Note that array bounds check is still enabled in release mode. There is a special build type for release builds with no bounds check. Similiarly, if you believe that contracts (pre -conditions, post-conditstions, invariants) are as valuable as bounds checks, then you should avoid release mode, where contracts are skipped. And use just "optimize" and "inline" instead. The "debug" is not about speed but about size: wouldn't you like to have debug symbols available to understand a backtrace in case of a fatal error?

I tend to use [`--build=release-debug`](https://dub.pm/dub-reference/buildtypes/#-buildrelease-debug). That way, we would have bound checks and symbol informations in case of exceptions. As far as I understood, we do not have pre-, post-conditions and invariant specifications either. So, we do not need to check them, isn't it?

Or should we define our own build type in _dub.json_? 
```
    "buildTypes": {
        "release-gamma": {
            "buildOptions": ["debugMode", "optimize", "inline", "debugInfo"]
        }
    }
```

kuniss commented:
>```
>     "buildTypes": {
>         "release-gamma": {
>             "buildOptions": ["debugMode", "optimize", "inline", "debugInfo"]
>         }
>     }
> ```

Tried this in my [PR detached branch](cross-compilation-pr-detached) and got a Linux binary that is bigger, the other binaries are smaller, see this [release page](https://github.com/kuniss/gamma/releases/tag/v0.3.0-delta.1).

I guess, we should run the [verification tests](https://github.com/kuniss/gamma/actions/runs/7439284147/job/20238999082) with the same build type and LCD compiled to be sure it works, isn't it?

kuniss commented:
> I guess, we should run the [verification tests](https://github.com/kuniss/gamma/actions/runs/7439284147/job/20238999082) with the same build type and LCD compiled to be sure it works, isn't it?

I tried this in this build run on the same branch as above: https://github.com/kuniss/gamma/actions/runs/7439461088
Works!

Now I'm thinking: Shouldn't we use LCD at _ci.yml_ too? I mean, at the end we are interested in checking LCD compiled gamma is working well. That it is working well with DMD, the "developer" should check locally, isn't it?

linkrope commented:
> Now I'm thinking: Shouldn't we use LCD at _ci.yml_ too? I mean, at the end we are interested in checking LCD compiled gamma is working well. That it is working well with DMD, the "developer" should check locally, isn't it?

This has to wait until "dmd" is no longer hard-coded in main.d? Otherwise, it could be tricky to install both compilers and use ldc to run the test, while dmd is used to compile the generated compiler.

kuniss commented:
I'm going to close this PR as it was substituted by PR #18 which has in general the same content but a newer origin base.
